### PR TITLE
Generate Lua bytecode for HTML5 bundles

### LINKED ISSUE #6752: Use Lua bytecode for non LuaJIT builds
**Task (REQUIRED):**
HTML5 builds do not generate Lua bytecode during the build process and instead build with Lua source code. 

**Expected outcome (REQUIRED):**
HTML5 builds should generate Lua bytecode.

### PR DESCRIPTION
**Pull request by britzl**
Builds for the HTML5 platform uses standard Lua 5.1.5 (while LuaJIT is used on all other platforms) and the Lua script files are included in the game archive as uncompiled text instead of bytecode. With this change the HTML5 platform also use bytecode (generated using the Lua 5.1.5 `luac` tool).

Fixes #6752

### CODE REVIEW

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Bob.java [hunk 1]
```diff
@@ -464,7 +464,8 @@ private static Options getCommandLineOptions() {
         addOption(options, null, "defoldsdk", true, "What version of the defold sdk (sha1) to use", true);
         addOption(options, null, "binary-output", true, "Location where built engine binary will be placed. Default is \"<build-output>/<platform>/\"", true);
 
-        addOption(options, null, "use-vanilla-lua", false, "Only ships vanilla source code (i.e. no byte code)", true);
+        addOption(options, null, "use-vanilla-lua", false, "DEPRECATED! Use --use-lua-source instead.", true);
+        addOption(options, null, "use-lua-source", false, "Use uncompressed and unencrypted Lua source code instead of byte code", true);
```
Comment (britzl) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Bob.java:468]: This is more representative of what actually happens. We bundle with Lua source code. It is not in any way connected to which Lua version we use.

**Comment by britzl** [comment]
The luac-32 and luac-64 executables for macOS, Windows and Linux are used to create 32 and 64 bit Lua bytecode (the same as we have for LuaJIT). The executables have been built from source on GitHub CI:

https://github.com/britzl/lua5.1.5/actions/runs/2574030871

The source has been modified to create luac executables that can generate 32 bit bytecode from a 64 bit host according to this lua-users posts: http://lua-users.org/lists/lua-l/2016-02/msg00176.html.

Are we ok with this solution?

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java [hunk 2]
```diff
@@ -253,25 +274,36 @@ public void build(Task<Void> task) throws CompileExceptionError, IOException {
...
         }
 
         LuaSource.Builder srcBuilder = LuaSource.newBuilder();
         srcBuilder.setFilename(task.input(0).getPath());
 
-        if (needsLuaSource.contains(project.getPlatform()) || use_vanilla_lua) {
+        if (use_lua_source) {
             srcBuilder.setScript(ByteString.copyFrom(script.getBytes()));
+        } else if (needsVanillaLua32.contains(project.getPlatform())) {
+            byte[] bytecode = constructLuaBytecode(task, "luac-32", script);
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java:300]: If we are not using luac-64, I don't think we should ship the executables?

**Review by JCash — DISMISSED** [review]
A small comment on distributing unused tools.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java [hunk 3]
```diff
@@ -253,25 +274,36 @@ public void build(Task<Void> task) throws CompileExceptionError, IOException {
...
-                String path = res.getAbsPath();
-                if(path.endsWith("luac") || path.endsWith("scriptc") || path.endsWith("gui_scriptc") || path.endsWith("render_scriptc")) {
+        boolean use_lua_source = this.project.option("use-lua-source", "false").equals("true");
+        
+        // set compression and encryption flags
+        // if the use-lua-source flag is set the project will use uncompressed plain text Lua script files
+        // if the use-lua-source flag is NOT set the project will use encrypted and possibly also compressed bytecode
+        for(IResource res : task.getOutputs()) {
+            String path = res.getAbsPath();
+            if(path.endsWith("luac") || path.endsWith("scriptc") || path.endsWith("gui_scriptc") || path.endsWith("render_scriptc")) {
```
[the remark this answers is not in the document]
Comment (britzl) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java:284]: Good question really. We had it in the old code as well. Surely the output of the LuaBuilder task is always one of those right?

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/GameProjectBuilder.java [hunk 4]
```diff
@@ -212,14 +209,15 @@ private void createArchive(Collection<String> resources, RandomAccessFile archiv
             }
         }
 
-        ArchiveBuilder archiveBuilder = new ArchiveBuilder(root, manifestBuilder, use_vanilla_lua ? false : true, resourcePadding);
-        boolean doCompress = project.getProjectProperties().getBooleanValue("project", "compress_archive", true);
+        ArchiveBuilder archiveBuilder = new ArchiveBuilder(root, manifestBuilder, resourcePadding);
 
+        boolean doCompress = project.getProjectProperties().getBooleanValue("project", "compress_archive", true);
         HashMap<String, EnumSet<Project.OutputFlags>> outputs = project.getOutputs();
```
[the remark this answers is not in the document]
Comment (britzl) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/GameProjectBuilder.java:215]: Good point. That might be something to look into for the future. I think what you suggest is a good solution.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java (hunk 2 above)
Comment (britzl) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java:300]: Removed luac-64

**Pull request merged**